Add documentation about rlp (#273)
zkworks-xyz/noir-ethereum-history-api: ethereum_history_api/circuits/lib/src/rlp/README.md

# RLP decoding

> This is a [Noir](https://noir-lang.org) library to decode [RLP](https://ethereum.org/en/developers/docs/data-structures-and-encoding/rlp/)

Reviewer: ```suggestion
This is a [Noir](https://noir-lang.org) library for [RLP](https://ethereum.org/en/developers/docs/data-structures-and-encoding/rlp/) decoding.
```

RLP encoding works with two types of items. A string & a list. Therefore this lib contains two main functions:

> - ```rust

Reviewer: I think this part is gonna look nicer if we don't write it as a list (without '-')

  pub fn decode_string<N>(data: Fragment) -> RlpFragment;
  ```
- ```rust
  pub fn decode_list<N, NUM_FIELDS>(data: Fragment) -> RlpList<NUM_FIELDS>;

>   ```

Reviewer: Maybe something like this would be more readable:
This library exports two functions:
- decode_string that takes an RLP-encoded string and returns an [RlpFragment](link) (link) - _short description of what RlpFragment is_
- decode_list that takes an RLP-encoded list and returns an [RlpList](link) (link) - _short description of what RlpList is_

Author: I think all these information are included in function headers

> - There is also a more performant special case version of `decode_list`, that can be used when we know that our input consists exclusively of strings of length <= 55 bytes.

Reviewer: ```suggestion
- The library also exports a more performant special case version of `decode_list`, that can be used when the input consists exclusively of strings of length <= 55 bytes:
```

  ```rust
  pub fn decode_list_of_small_strings<N, NUM_FIELDS>(data: Fragment) -> RlpList<NUM_FIELDS>;
  ```

## Example

```rust

> let rlp = [0x82, 0x10, 0x00]; // "0x1000"
> decode_string(rlp.into());

Reviewer: ```suggestion
let rlp = Fragment::from_array([0x82, 0x10, 0x00]); // "0x1000"
decode_string(rlp);
```
This will need to be changed depending on the output of [this PR](https://github.com/vlayer-xyz/noir-ethereum-history-api/pull/275), but `.into()` very likely won't work. `Fragment::from_array()` is more likely to work

// {offset: 1, length: 2, data_type: STRING}

let rlp = [0xc2, 0x80, 0x10]; // ["", "0x10"]

> let decoded = decode_list(rlp.into());

Reviewer: As above

// [{offset: 2, length: 0, data_type: STRING}, {offset: 2, length: 1, data_type: STRING}]

> // Or using our helper assert methods:

Reviewer: ```suggestion
// Or using helper assert methods:
```

Reviewer: Also I think these 3 lines worsen the readability as it is not clear how `assert_eq_empty_string` and `assert_eq_u64` work

Author: But we wanted to show these functions.

decoded.get(0).assert_eq_empty_string("Field 0");
decoded.get(1).assert_eq_u64("Field 1", rlp, 0x10);
```

> ## What is Fragment

Reviewer: ```suggestion
**What is a Fragment?**
```
It is a subpart of the example so it should be smaller

[Fragment](../misc/README.md) is a dynamically-sized view into a comptime-sized sequence of elements.  

> In the example above, Into trait converts array into Fragment.

Reviewer: ```suggestion
```
This won't be needed if we use `Fragment::from_array()`

# RLP types

## RlpFragment

```Rust
struct RlpFragment {
    offset: u64,
    length: u64,
    data_type: u64 // STRING or LIST
}
```

> It exists only in connection with the particular array of RLP encoding as it doesn't provide data itself. Ideally we would want to use only `Fragment` which includes data. Current Noir state makes using data, in each case when we use RlpFragment, much too inefficient so we use RlpFragment for performance reasons.

Reviewer: ```suggestion
RlpFragment exists only in connection with a particular array of RLP encoding as it doesn't store data itself. _Here a little about what offset, length and data_type in RlpFragment are._ Ideally we would want to use `Fragment` instead of `RlpFragment` but current Noir state makes it too inefficient.
```

Author: I don't think we need to explain what offset and length is

> There are several comparison methods in RlpFragment implementation. They check if in given RLP encoding array, at position set by this fragment, there is given data. Each of these function is made to receive data as different type to simplify its usage:

Reviewer: ```suggestion
There are several comparison methods in RlpFragment implementation. They check if in given RLP encoding array, at offset specified by the fragment, lays given data. Each of these functions is made to receive data as different type to simplify its usage:
```

Reviewer: I don't understand the last sentence, maybe it could be rewritten?

Author: I rewrote it a bit differently...

> There are several comparison methods in RlpFragment implementation. They check if in given RLP encoding array, at position set by this fragment, there is given data. Each of these function is made to receive data as different type to simplify its usage:

Reviewer: I think these comparison methods are more internal (for our team's use) than external, so I think they should be mentioned later

Author: I think they are more external than internal

- `assert_eq_bytes`
- `assert_eq_bounded_vec`
- `assert_empty_string`
- `assert_eq_u1`
- `assert_eq_u8`
- `assert_eq_u32`
- `assert_eq_u64`
- `assert_eq_u128`
- `assert_eq_address`
- `assert_eq_bytes32`

## RlpList

```Rust
type RlpList<MAX_FIELDS> = BoundedVec<RlpFragment, MAX_FIELDS>;

> ```

Reviewer: I would add a sentence explaining what `RlpList` is

Author: I think its definition is pretty self-explanatory.

> # Interface

Reviewer: I think all of the information in Interface section should not be here in this separate section, but at the beginning, when we first introduce the functions. There's not much new information here, so I think placing it at the end creates confusion.

These functions receive RLP encoding as `Fragment` type and return information about offset, length and data type of data encoded.  
**Important:** Returned offset doesn't include Fragment's offset.  

> Data should contain whole element with its RLP header.

Reviewer: The last sentence isn't too clear, and also I'm not sure if it's needed, but if it is, maybe it could be rewritten

Author: I think it's needed
